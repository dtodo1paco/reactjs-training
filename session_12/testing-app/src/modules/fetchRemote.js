
export const add = (a,b) => a+b;

export const fetchData = async(url) => {
  const resp = await fetch(url);
  const parsedResponse = await resp.json();
  return parsedResponse;
}

