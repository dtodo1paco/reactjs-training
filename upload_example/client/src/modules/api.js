export const BACKEND_API = "/api/v1";
export const BACKEND_API_UPLOAD = `${BACKEND_API}/upload`;

export const postData = async(url, formData) => {
  const resp = await fetch(url, {
    method: 'POST',
    body: formData,
  });
  const parsedResponse = await resp.json();
  return parsedResponse;
}

export const buildFormData = files => {
  const formData = new FormData()
  files.forEach((file, i) => {
    formData.append(i, file)
  })
  return formData;
}

export const uploadFiles = async(files) => {
  const formData = buildFormData(files);
  return await postData(BACKEND_API_UPLOAD, formData);
}


