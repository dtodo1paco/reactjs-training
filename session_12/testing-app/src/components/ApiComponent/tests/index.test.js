import React from 'react';
import { shallow, mount } from 'enzyme';
import Component from '../index';

import { fetchData } from '../../../modules/fetchRemote';

jest.mock("../../../modules/fetchRemote"); // same for all exports
//
// mock responses
//
fetchData
  .mockReturnValueOnce({name:"dtodo1paco", counter: 1})
  .mockReturnValueOnce({name:"dtodo1paco", counter: 2})
  .mockReturnValue({name:"dtodo1paco", counter: 3});


describe("ApiComponent basics", () => {
  it("should render correctly", () => {
    const wrapper = shallow(<Component url="test"/>);
  });
  it("should render initial layout", () => {
    const wrapper = shallow(<Component url="test"/>);
    expect(wrapper.getElements()).toMatchSnapshot();
  });
  it("should NOT initialize component state", () => {
    const wrapper = shallow(<Component url="test"/>);
    expect(wrapper.state('response')).not.toBeDefined();
  });

  it('receive props correctly', () => {
    const wrapper = shallow(
      <Component url="test" />
    );
    expect(wrapper.instance().props['url']).toEqual("test");
  });
});

describe("ApiComponent event handling", () => {
  it('call API when button clicked', () => {
    const wrapper = shallow(
      <Component url="testURL"/>
    );
    const button = wrapper.find('button');
    button.simulate('click');
    expect(fetchData).toHaveBeenCalled();

  });
  it('call API when button clicked with correct params', () => {
    const wrapper = shallow(
      <Component url="testURL"/>
    );
    const button = wrapper.find('button');
    button.simulate('click');
    expect(fetchData).toHaveBeenCalledWith("testURL")
  });
});

describe("ApiComponent state update", () => {

  //
  // async method, needs await
  //
  it('call API when button clicked set state to response', async () => {
    const wrapper = shallow(
      <Component url="testURL"/>
    );
    const button = wrapper.find('button');
    //
    // IMPORTANT: buttton click will wait API to change state
    //
    await button.simulate('click');
    expect(wrapper.state('response')).toBeDefined();
  });

  it('call API when button clicked set state to response', async () => {
    const wrapper = shallow(
      <Component url="testURL"/>
    );
    const button = wrapper.find('button');
    //
    // IMPORTANT: buttton click will wait API to change state
    //
    await button.simulate('click');
    expect(wrapper.state('response').counter).toEqual(3);
  });
});
