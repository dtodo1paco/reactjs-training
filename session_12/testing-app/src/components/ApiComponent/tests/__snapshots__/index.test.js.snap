// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiComponent basics should render initial layout 1`] = `
Array [
  <div
    className="myComponentClass"
  >
    Component is here 
    <button
      onClick={[Function]}
    >
      Click me to call an API
    </button>
  </div>,
]
`;
